py/server/deephaven/_udf.py

```diff
@@ -559,53 +560,57 @@ def _wrapper(*args, **kwargs):
                 if return_array:
                     return dtypes.array(ret_dtype, ret)
                 else:
-                    return p_sig.ret_annotation.ret_converter(ret) if p_sig.ret_annotation.ret_converter else ret
+                    return ret_converter(ret) if ret_converter else ret
 
             return _wrapper
         else:  # for vectorization
+            if encoded_arg_types and len(encoded_arg_types.split(",")) != len(p_sig.params):
+                raise ValueError(
+                    f"The number of arguments doesn't match the function ({p_sig.fn.__name__}) signature. "
+                    f"{len(encoded_arg_types.split(','))}, {p_sig.encoded}")
+
             def _vectorization_wrapper(*args):
-                if len(args) != len(p_sig.params) + 2:
-                    raise ValueError(
-                        f"The number of arguments doesn't match the function ({p_sig.fn.__name__}) signature. "
-                        f"{len(args) - 2}, {p_sig.encoded}")
                 if args[0] <= 0:
                     raise ValueError(
                         f"The chunk size argument must be a positive integer for vectorized function ("
                         f"{p_sig.fn.__name__}). {args[0]}")
 
                 chunk_size = args[0]
                 chunk_result = args[1]
-                if args[2:]:
+                if args[2:] and arg_conv_needed:  # if there are arguments
```

Comment:
Still branchy

Reply:
also concerned that `args[2:]` is going to be slow.

Would we be better off to change `args` to `Tuple[size, result, args]` to avoid these list operations?